src/openforms/prefill/service.py

```diff
@@ -0,0 +1,188 @@
+is responsible for exposing which attributes/data fragments are available, and for performing
+the actual look-up. Plugins receive the :class:`openforms.submissions.models.Submission` 
+instance that represents the current form session of an end-user.
+
+Prefill values are embedded as default values for form fields, dynamically for every
+user session using the component rewrite functionality in the serializers.
+
+So, to recap:
+
+1. Plugins are defined and registered
+2. When editing form definitions in the admin, content editors can opt-in to pre-fill
+   functionality. They select the desired plugin, and then the desired attribute from
+   that plugin.
+3. Content editors can also define a user-defined variable and configure the plugin and
+   the necessary options by selecting the desired choices for the ``prefill_options``.
+4. End-user starts the form and logs in, thereby creating a session/``Submission``
+5. The submission-specific form definition configuration is enhanced with the pre-filled
+   form field default values.
+"""
+
+import logging
+from collections import defaultdict
+
+import elasticapm
+
+from openforms.formio.service import FormioConfigurationWrapper
+from openforms.submissions.models import Submission
+from openforms.submissions.models.submission_value_variable import (
+    SubmissionValueVariable,
+)
+from openforms.typing import JSONEncodable
+from openforms.variables.constants import FormVariableSources
+
+from .registry import Registry
+from .sources import (
+    fetch_prefill_values_from_attribute,
+    fetch_prefill_values_from_options,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def inject_prefill(
+    configuration_wrapper: FormioConfigurationWrapper, submission: Submission
+) -> None:
+    """
+    Mutates each component found in configuration according to the prefilled values.
+
+    :param configuration_wrapper: The Formiojs JSON schema wrapper describing an entire
+    form or an individual component within the form.
+    :param submission: The :class:`openforms.submissions.models.Submission` instance
+    that holds the values of the prefill data. The prefill data was fetched earlier,
+    see :func:`prefill_variables`.
+
+    The prefill values are looped over by key: value, and for each value the matching
+    component is looked up to normalize it in the context of the component.
+    """
+
+    from openforms.formio.service import normalize_value_for_component
+
+    prefilled_data = submission.get_prefilled_data()
+    for key, prefill_value in prefilled_data.items():
+        try:
+            component = configuration_wrapper[key]
+        except KeyError:
+            # The component to prefill is not in this step
+            continue
+
+        if not (prefill := component.get("prefill")):
+            continue
+        if not prefill.get("plugin"):
+            continue
+        if not prefill.get("attribute"):
+            continue
```

Comment:
Lines were not covered before the code was moved, accepting this as-is and I'm ignoring the codecov annotations.